### PR TITLE
[SPARK-45639][SQL][PYTHON] Support loading Python data sources in DataFrameReader

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR supports `spark.read.format(...).load()` for Python data sources.

After this PR, users can use a Python data source directly like this:
```python
from pyspark.sql.datasource import DataSource, DataSourceReader

class MyReader(DataSourceReader):
    def read(self, partition):
        yield (0, 1)

class MyDataSource(DataSource):
    @classmethod
    def name(cls):
        return "my-source"

    def schema(self):
        return "id INT, value INT"
    
    def reader(self, schema):
        return MyReader()

spark.dataSource.register(MyDataSource)

df = spark.read.format("my-source").load()
df.show()
+---+-----+
| id|value|
+---+-----+
|  0|    1|
+---+-----+
```

### Why are the changes needed?

To support Python data sources.

### Does this PR introduce _any_ user-facing change?

Yes. After this PR, users can load a custom Python data source using `spark.read.format(...).load()`.

### How was this patch tested?

New unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 1]
```diff
@@ -110,11 +111,13 @@ private[sql] class SharedState(
   /**
    * A data source manager shared by all sessions.
    */
+  @Unstable
   lazy val dataSourceManager = new DataSourceManager()
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala:115]: I think we actually don't need to expose this and below as an API?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceManager.scala [hunk 2]
```diff
@@ -20,35 +20,60 @@ package org.apache.spark.sql.execution.datasources
...
+import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap
 import org.apache.spark.sql.errors.QueryCompilationErrors
 import org.apache.spark.sql.types.StructType
-import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
+/**
+ * A manager for user-defined data sources. It is used to register and lookup data sources by
+ * their short names or fully qualified names.
+ */
+@Unstable
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceManager.scala:34]: This too. Let's just remove this `@Unstable`.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -231,6 +231,7 @@ class SparkSession private(
   /**
    * A collection of methods for registering user-defined data sources.
    */
+  @Unstable
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:234]: We should either remove `@Unstable` or `private[sql]` if we want this as an API.

Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:234]: I'll make it `private[sql]` for now before we decide to expose this as an API.

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 4]
```diff
@@ -208,10 +209,45 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       throw QueryCompilationErrors.pathOptionNotSetCorrectlyWhenReadingError()
     }
 
-    DataSource.lookupDataSourceV2(source, sparkSession.sessionState.conf).flatMap { provider =>
-      DataSourceV2Utils.loadV2Source(sparkSession, provider, userSpecifiedSchema, extraOptions,
-        source, paths: _*)
-    }.getOrElse(loadV1Source(paths: _*))
+    val isUserDefinedDataSource =
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:212]: @cloud-fan @allisonwang-db do we want to support this datasource via `USING` syntax unlike DSv2, right?

If that's the case, the logics of loading DataSource has to be within `DataSource.lookupDataSource` and/or `DataSource.providingInstance`. I don't think we should mix the logics here with DSv2.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:212]: Let's at least separate the logic into a separate function if possible.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:212]: Unfortunately DS v2 `TableProvider` does not support USING yet. That's why the code is a bit messy here as it's not shared with the SQL USING path. We should support it though...

**Comment by HyukjinKwon** [comment]
Merged to master.

I had some offline discussion, and I will follow up on my own.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by allisonwang-db** [comment]
Thanks @HyukjinKwon!

**Comment by HyukjinKwon** [comment]
Made a POC (or draft?) PTAL: https://github.com/apache/spark/pull/43784

**Pull request closed without merging**